src/main/java/de/tum/cit/aet/artemis/core/service/LLMTokenUsageService.java

```diff
@@ -0,0 +1,134 @@
+     *
+     * @param llmRequests     List of LLM requests
+     * @param serviceType     type of the LLM service
+     * @param builderFunction of type Function<IrisTokenUsageBuilder, IrisTokenUsageBuilder> using IrisTokenUsageBuilder
+     * @return saved LLMTokenUsage as a List
+     */
+    // TODO: this should ideally be done Async
+    public LLMTokenUsageTrace saveLLMTokenUsage(List<LLMRequest> llmRequests, LLMServiceType serviceType, Function<LLMTokenUsageBuilder, LLMTokenUsageBuilder> builderFunction) {
+        LLMTokenUsageTrace llmTokenUsageTrace = new LLMTokenUsageTrace();
+        llmTokenUsageTrace.setServiceType(serviceType);
+
+        LLMTokenUsageBuilder builder = builderFunction.apply(new LLMTokenUsageBuilder());
+        builder.getIrisMessageID().ifPresent(llmTokenUsageTrace::setIrisMessageId);
+        builder.getUser().ifPresent(user -> llmTokenUsageTrace.setUserId(user.getId()));
+        builder.getExercise().ifPresent(exercise -> llmTokenUsageTrace.setExerciseId(exercise.getId()));
+        builder.getCourse().ifPresent(course -> llmTokenUsageTrace.setCourseId(course.getId()));
+
+        Set<LLMTokenUsageRequest> llmRequestsSet = llmTokenUsageTrace.getLLMRequests();
+        setLLMTokenUsageRequests(llmRequests, llmTokenUsageTrace, llmRequestsSet);
+        return llmTokenUsageTraceRepository.save(llmTokenUsageTrace);
+    }
+
+    private void setLLMTokenUsageRequests(List<LLMRequest> llmRequests, LLMTokenUsageTrace llmTokenUsageTrace, Set<LLMTokenUsageRequest> llmRequestsSet) {
+        for (LLMRequest llmRequest : llmRequests) {
+            LLMTokenUsageRequest llmTokenUsageRequest = new LLMTokenUsageRequest();
+            llmTokenUsageRequest.setModel(llmRequest.model());
+            llmTokenUsageRequest.setNumInputTokens(llmRequest.numInputTokens());
+            llmTokenUsageRequest.setNumOutputTokens(llmRequest.numOutputTokens());
+            llmTokenUsageRequest.setCostPerMillionInputTokens(llmRequest.costPerMillionInputToken());
+            llmTokenUsageRequest.setCostPerMillionOutputTokens(llmRequest.costPerMillionOutputToken());
+            llmTokenUsageRequest.setServicePipelineId(llmRequest.pipelineId());
+            llmTokenUsageRequest.setTrace(llmTokenUsageTrace);
+            llmRequestsSet.add(llmTokenUsageRequest);
+        }
+    }
+
+    // TODO: this should ideally be done Async
+    public void appendRequestsToTrace(List<LLMRequest> requests, LLMTokenUsageTrace trace) {
+        Set<LLMTokenUsageRequest> llmRequestsSet = new HashSet<>();
+        setLLMTokenUsageRequests(requests, trace, llmRequestsSet);
+        llmTokenUsageRequestRepository.saveAll(llmRequestsSet);
+    }
+
+    /**
+     * Class LLMTokenUsageBuilder to be used for saveLLMTokenUsage()
+     */
+    public static class LLMTokenUsageBuilder {
+
+        private Optional<Course> course = Optional.empty();
+
+        private Optional<Long> irisMessageID = Optional.empty();
```

Comment:
This is inconsistent. This is the only builder attribute which is an ID, while the rest are domain objects. I would suggest choosing one or the other. Only the ID is really necessary when the builder is consumed, so I would suggest switching to IDs for all. This also makes the builder more flexible for the caller, since only the ID needs to be known (otherwise the caller has to do a database call just for the domain object to be thrown away again anyway - which I see that you actually had to do in `IrisCompetencyGenerationService`!).

Reply:
That is an excellent point! @FelixTJDietrich, That change would also affect you, what do you think about that?